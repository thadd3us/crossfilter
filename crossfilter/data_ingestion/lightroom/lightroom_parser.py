"""Parser for Adobe Lightroom catalog files."""

import logging
import os
import tempfile
import zipfile
from pathlib import Path
from typing import Dict, List, Optional, Set

import pandas as pd
from pydantic import BaseModel
from sqlalchemy import create_engine, text

from crossfilter.core.schema import DataType, SchemaColumns
from crossfilter.data_ingestion.sqlite_utils import query_sqlite_to_dataframe

logger = logging.getLogger(__name__)


class LightroomParserConfig(BaseModel):
    """Configuration for Lightroom parser."""

    ignore_collections: Set[str] = {"quick collection"}
    include_metadata: bool = True
    include_keywords: bool = True
    include_collections: bool = True


def is_zip_file(path: Path) -> bool:
    """Check if a path is a zip file."""
    return path.suffix.lower() == ".zip"


def extract_catalog_from_zip(zip_path: Path) -> Path:
    """Extract Lightroom catalog from a zip file to a temporary directory.

    WARNING: The returned path points to a file in a temporary directory.
    The caller is responsible for cleaning up the temporary directory when done.
    """
    logger.info(f"Extracting catalog from zip file: {zip_path}")

    # Create a temporary directory for extraction
    temp_dir = Path(tempfile.mkdtemp(prefix="lightroom_catalog_"))

    try:
        with zipfile.ZipFile(zip_path, "r") as zip_ref:
            zip_ref.extractall(temp_dir)

        # Find the .lrcat file
        lrcat_files = list(temp_dir.glob("*.lrcat"))
        if not lrcat_files:
            raise ValueError(f"No .lrcat file found in {zip_path}")

        if len(lrcat_files) > 1:
            logger.warning(
                f"Multiple .lrcat files found in {zip_path}, using the first one"
            )

        catalog_path = lrcat_files[0]
        logger.info(f"Extracted catalog to: {catalog_path}")
        return catalog_path

    except Exception as e:
        # Clean up temp directory if extraction fails
        import shutil

        shutil.rmtree(temp_dir, ignore_errors=True)
        raise e


def determine_file_type(file_format: str) -> DataType:
    """Determine if a file is a photo or video based on Lightroom's fileFormat column."""
    if not file_format:
        return DataType.PHOTO  # Default to photo

    # Lightroom's fileFormat values for video files
    video_formats = {
        "MP4",
        "MOV",
        "AVI",
        "MKV",
        "WMV",
        "FLV",
        "WEBM",
        "M4V",
        "3GP",
        "3G2",
        "MTS",
        "M2TS",
        "TS",
        "MXF",
        "DV",
        "F4V",
        "ASF",
        "RM",
        "RMVB",
        "VOB",
        "OGV",
    }

    if file_format.upper() in video_formats:
        return DataType.VIDEO
    else:
        return DataType.PHOTO


def build_file_path(
    root_path: Optional[str], path_from_root: Optional[str], filename: Optional[str]
) -> Optional[str]:
    """Build full file path from Lightroom components."""
    if not filename:
        return None

    parts = []
    if root_path:
        parts.append(root_path)
    if path_from_root:
        parts.append(path_from_root)
    parts.append(filename)

    return os.path.join(*parts)


def parse_lightroom_catalog(
    catalog_path: Path, config: LightroomParserConfig = None
) -> pd.DataFrame:
    """Parse a Lightroom catalog file and return a DataFrame with the target schema."""
    if config is None:
        config = LightroomParserConfig()

    original_path = catalog_path
    extracted_temp_dir = None

    try:
        # Handle zip files
        if is_zip_file(catalog_path):
            catalog_path = extract_catalog_from_zip(catalog_path)
            # Remember the temp directory for cleanup
            extracted_temp_dir = catalog_path.parent

        if not catalog_path.exists():
            raise FileNotFoundError(f"Lightroom catalog not found: {catalog_path}")

        logger.info(f"Parsing Lightroom catalog: {catalog_path}")

        # Main query to get image data with basic joins
        # Start with a simpler query to ensure compatibility
        main_query = """
        SELECT DISTINCT
            -- Core Adobe_images fields
            Adobe_images.id_global,
            Adobe_images.rootFile,
            Adobe_images.captureTime,
            Adobe_images.rating,
            Adobe_images.colorLabels,
            Adobe_images.fileFormat,
            Adobe_images.originalCaptureTime,
            Adobe_images.pick,
            Adobe_images.touchTime,
            
            -- File information
            AgLibraryFile.idx_filename,
            AgLibraryFile.importHash,
            AgLibraryFile.baseName,
            AgLibraryFile.extension,
            
            -- Folder information 
            AgLibraryFolder.pathFromRoot,
            AgLibraryRootFolder.absolutePath,
            
            -- GPS and EXIF metadata (only fields that commonly exist)
            AgHarvestedExifMetadata.gpsLatitude,
            AgHarvestedExifMetadata.gpsLongitude,
            AgHarvestedExifMetadata.focalLength,
            AgHarvestedExifMetadata.aperture,
            AgHarvestedExifMetadata.flashFired,
            
            -- IPTC metadata (basic fields)
            AgLibraryIPTC.caption
            
        FROM Adobe_images
        LEFT JOIN AgLibraryFile ON AgLibraryFile.id_local = Adobe_images.rootFile
        LEFT JOIN AgLibraryFolder ON AgLibraryFolder.id_local = AgLibraryFile.folder
        LEFT JOIN AgLibraryRootFolder ON AgLibraryRootFolder.id_local = AgLibraryFolder.rootFolder
        LEFT JOIN AgHarvestedExifMetadata ON AgHarvestedExifMetadata.image = Adobe_images.id_local
        LEFT JOIN AgLibraryIPTC ON AgLibraryIPTC.image = Adobe_images.id_local
        """

        # Query the main data
        df = query_sqlite_to_dataframe(catalog_path, main_query)

        if df.empty:
            logger.warning("No images found in Lightroom catalog")
            return pd.DataFrame(
                columns=[
                    SchemaColumns.UUID_STRING,
                    SchemaColumns.DATA_TYPE,
                    SchemaColumns.NAME,
                    SchemaColumns.CAPTION,
                    SchemaColumns.SOURCE_FILE,
                    SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
                    SchemaColumns.TIMESTAMP_UTC,
                    SchemaColumns.GPS_LATITUDE,
                    SchemaColumns.GPS_LONGITUDE,
                    SchemaColumns.RATING_0_TO_5,
                    SchemaColumns.SIZE_IN_BYTES,
                ]
            )

        logger.info(f"Found {len(df)} images in Lightroom catalog")

        # Map to target schema columns
        result_df = pd.DataFrame()

        # Required schema columns
        result_df[SchemaColumns.UUID_STRING] = df["id_global"].astype(str)
        result_df[SchemaColumns.DATA_TYPE] = df["fileFormat"].apply(determine_file_type)
        result_df[SchemaColumns.NAME] = df["idx_filename"]  # Use filename as name
        result_df[SchemaColumns.CAPTION] = df["caption"]

        # Build source file path
        result_df[SchemaColumns.SOURCE_FILE] = df.apply(
            lambda row: build_file_path(
                row.get("absolutePath"),
                row.get("pathFromRoot"),
                row.get("idx_filename"),
            ),
            axis=1,
        )

        # Handle timestamps
        result_df[SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE] = df[
            "captureTime"
        ]  # THAD: Can we actually get tz-aware timestamps?
        result_df[SchemaColumns.TIMESTAMP_UTC] = df["captureTime"]
        if result_df[SchemaColumns.TIMESTAMP_UTC].isna().any():
            logger.warning(
                f"Found {result_df[SchemaColumns.TIMESTAMP_UTC].isna().sum()} missing timestamps"
            )

        # GPS coordinates
        result_df[SchemaColumns.GPS_LATITUDE] = pd.to_numeric(
            df["gpsLatitude"], errors="coerce"
        )
        result_df[SchemaColumns.GPS_LONGITUDE] = pd.to_numeric(
            df["gpsLongitude"], errors="coerce"
        )

        # Rating (0-5)
        result_df[SchemaColumns.RATING_0_TO_5] = (
            pd.to_numeric(df["rating"], errors="coerce").fillna(0).astype("Int64")
        )

        # File size (not available in simplified query, set to None)
        result_df[SchemaColumns.SIZE_IN_BYTES] = None

        # Add useful metadata columns as extras (not part of required schema)
        if config.include_metadata:
            # EXIF data (available from our simplified query)
            result_df["focal_length"] = pd.to_numeric(
                df["focalLength"], errors="coerce"
            )
            result_df["aperture"] = pd.to_numeric(df["aperture"], errors="coerce")
            result_df["flash_fired"] = df["flashFired"]

            # File metadata
            result_df["import_hash"] = df["importHash"]
            result_df["file_format"] = df["fileFormat"]
            result_df["base_name"] = df["baseName"]
            result_df["extension"] = df["extension"]

            # Lightroom-specific fields
            result_df["color_labels"] = df["colorLabels"]
            result_df["pick"] = df["pick"]
            result_df["touch_time"] = df["touchTime"]
            result_df["original_capture_time"] = df["originalCaptureTime"]

        # Add keywords if requested
        if config.include_keywords:
            keywords_df = _get_keywords_for_images(
                catalog_path, result_df[SchemaColumns.UUID_STRING].tolist()
            )
            result_df = result_df.merge(
                keywords_df, on=SchemaColumns.UUID_STRING, how="left"
            )
            # Convert keywords list to JSON string for SQLite storage
            if "keywords" in result_df.columns:
                result_df["keywords"] = result_df["keywords"].apply(
                    lambda x: str(x) if x is not None else None
                )

        # Add collections if requested
        if config.include_collections:
            collections_df = _get_collections_for_images(
                catalog_path,
                result_df[SchemaColumns.UUID_STRING].tolist(),
                config.ignore_collections,
            )
            result_df = result_df.merge(
                collections_df, on=SchemaColumns.UUID_STRING, how="left"
            )
            # Convert collections list to JSON string for SQLite storage
            if "collections" in result_df.columns:
                result_df["collections"] = result_df["collections"].apply(
                    lambda x: str(x) if x is not None else None
                )

        logger.info(f"Parsed {len(result_df)} records from Lightroom catalog")

        return result_df

    finally:
        # Clean up temporary directory if we extracted from zip
        if extracted_temp_dir:
            try:
                import shutil

                shutil.rmtree(extracted_temp_dir, ignore_errors=True)
                logger.debug(f"Cleaned up temporary directory {extracted_temp_dir}")
            except Exception as e:
                logger.warning(
                    f"Failed to clean up temporary directory {extracted_temp_dir}: {e}"
                )


def _get_keywords_for_images(catalog_path: Path, image_ids: List[str]) -> pd.DataFrame:
    """Get keywords for the specified image IDs."""
    if not image_ids:
        return pd.DataFrame(columns=[SchemaColumns.UUID_STRING, "keywords"])

    # Convert image_ids to a format suitable for SQL IN clause
    ids_str = ",".join(f"'{id_}'" for id_ in image_ids)

    keywords_query = f"""
    SELECT 
        Adobe_images.id_global,
        AgLibraryKeyword.name as keyword_name
    FROM AgLibraryKeywordImage
    LEFT JOIN Adobe_images ON Adobe_images.id_local = AgLibraryKeywordImage.image
    LEFT JOIN AgLibraryKeyword ON AgLibraryKeyword.id_local = AgLibraryKeywordImage.tag
    WHERE Adobe_images.id_global IN ({ids_str})
    AND AgLibraryKeyword.name IS NOT NULL
    """

    try:
        keywords_df = query_sqlite_to_dataframe(catalog_path, keywords_query)
        if not keywords_df.empty:
            # Group keywords by image ID
            grouped = (
                keywords_df.groupby("id_global")["keyword_name"]
                .apply(lambda x: sorted(x.dropna().tolist()))
                .reset_index()
            )
            grouped.columns = [SchemaColumns.UUID_STRING, "keywords"]
            grouped[SchemaColumns.UUID_STRING] = grouped[
                SchemaColumns.UUID_STRING
            ].astype(str)
            return grouped
    except Exception as e:
        logger.warning(f"Failed to get keywords: {e}")

    return pd.DataFrame(columns=[SchemaColumns.UUID_STRING, "keywords"])


def _get_collections_for_images(
    catalog_path: Path, image_ids: List[str], ignore_collections: Set[str]
) -> pd.DataFrame:
    """Get collections for the specified image IDs."""
    if not image_ids:
        return pd.DataFrame(columns=[SchemaColumns.UUID_STRING, "collections"])

    # Convert image_ids to a format suitable for SQL IN clause
    ids_str = ",".join(f"'{id_}'" for id_ in image_ids)

    # Convert ignore_collections to SQL format (case-insensitive)
    ignore_str = ",".join(f"'{col.lower()}'" for col in ignore_collections)
    ignore_clause = (
        f"AND LOWER(AgLibraryCollection.name) NOT IN ({ignore_str})"
        if ignore_collections
        else ""
    )

    collections_query = f"""
    SELECT 
        Adobe_images.id_global,
        AgLibraryCollection.name as collection_name
    FROM AgLibraryCollectionImage
    LEFT JOIN Adobe_images ON Adobe_images.id_local = AgLibraryCollectionImage.image
    LEFT JOIN AgLibraryCollection ON AgLibraryCollection.id_local = AgLibraryCollectionImage.collection
    WHERE Adobe_images.id_global IN ({ids_str})
    AND AgLibraryCollection.name IS NOT NULL
    {ignore_clause}
    """

    try:
        collections_df = query_sqlite_to_dataframe(catalog_path, collections_query)
        if not collections_df.empty:
            # Group collections by image ID
            grouped = (
                collections_df.groupby("id_global")["collection_name"]
                .apply(lambda x: sorted(x.dropna().tolist()))
                .reset_index()
            )
            grouped.columns = [SchemaColumns.UUID_STRING, "collections"]
            grouped[SchemaColumns.UUID_STRING] = grouped[
                SchemaColumns.UUID_STRING
            ].astype(str)
            return grouped
    except Exception as e:
        logger.warning(f"Failed to get collections: {e}")

    return pd.DataFrame(columns=[SchemaColumns.UUID_STRING, "collections"])


def load_lightroom_catalog_to_df(
    catalog_path: Path, config: LightroomParserConfig = None
) -> pd.DataFrame:
    """Load a Lightroom catalog and return a DataFrame with the target schema.

    This is the main entry point for loading Lightroom catalogs.
    """
    try:
        return parse_lightroom_catalog(catalog_path, config)
    except Exception as e:
        logger.error(f"Failed to load Lightroom catalog {catalog_path}: {e}")
        # Return empty DataFrame with correct schema on error
        return pd.DataFrame(
            columns=[
                SchemaColumns.UUID_STRING,
                SchemaColumns.DATA_TYPE,
                SchemaColumns.NAME,
                SchemaColumns.CAPTION,
                SchemaColumns.SOURCE_FILE,
                SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
                SchemaColumns.TIMESTAMP_UTC,
                SchemaColumns.GPS_LATITUDE,
                SchemaColumns.GPS_LONGITUDE,
                SchemaColumns.RATING_0_TO_5,
                SchemaColumns.SIZE_IN_BYTES,
            ]
        )
