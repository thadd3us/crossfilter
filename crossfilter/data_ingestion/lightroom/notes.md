* Sqlite can store captureTime with a timezone, but lightroom mostly doesn't do it.  Only FastFoto images have this set.
