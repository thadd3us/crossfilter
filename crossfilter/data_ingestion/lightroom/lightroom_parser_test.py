"""Tests for Lightroom catalog parser."""

import tempfile
import zipfile
from pathlib import Path
from typing import Set

import pandas as pd
import pytest

from crossfilter.core.schema import DataType, SchemaColumns
from crossfilter.data_ingestion.lightroom.lightroom_parser import (
    LightroomParserConfig,
    build_file_path,
    determine_file_type,
    extract_catalog_from_zip,
    is_zip_file,
    load_lightroom_catalog_to_df,
    parse_lightroom_catalog,
)


def test_determine_file_type() -> None:
    """Test determining file type from Lightroom's fileFormat column."""
    # Photo formats (Lightroom fileFormat values)
    assert determine_file_type("JPEG") == DataType.PHOTO
    assert determine_file_type("DNG") == DataType.PHOTO
    assert determine_file_type("TIFF") == DataType.PHOTO
    assert determine_file_type("PNG") == DataType.PHOTO
    assert determine_file_type("RAW") == DataType.PHOTO

    # Video formats (Lightroom fileFormat values)
    assert determine_file_type("MP4") == DataType.VIDEO
    assert determine_file_type("mp4") == DataType.VIDEO
    assert determine_file_type("MOV") == DataType.VIDEO
    assert determine_file_type("AVI") == DataType.VIDEO
    assert determine_file_type("MKV") == DataType.VIDEO
    assert determine_file_type("WMV") == DataType.VIDEO
    assert determine_file_type("M4V") == DataType.VIDEO

    # Edge cases
    assert determine_file_type("") == DataType.PHOTO  # Default to photo
    assert determine_file_type(None) == DataType.PHOTO
    assert determine_file_type("UNKNOWN") == DataType.PHOTO


def test_build_file_path() -> None:
    """Test building file paths from Lightroom components."""
    # Complete path
    result = build_file_path("/root/path", "subfolder", "image.jpg")
    assert result == "/root/path/subfolder/image.jpg"

    # Missing path_from_root
    result = build_file_path("/root/path", None, "image.jpg")
    assert result == "/root/path/image.jpg"

    # Missing root_path
    result = build_file_path(None, "subfolder", "image.jpg")
    assert result == "subfolder/image.jpg"

    # Only filename
    result = build_file_path(None, None, "image.jpg")
    assert result == "image.jpg"

    # Missing filename
    result = build_file_path("/root/path", "subfolder", None)
    assert result is None

    result = build_file_path("/root/path", "subfolder", "")
    assert result is None


def test_is_zip_file() -> None:
    """Test zip file detection."""
    assert is_zip_file(Path("catalog.zip"))
    assert is_zip_file(Path("catalog.ZIP"))
    assert not is_zip_file(Path("catalog.lrcat"))
    assert not is_zip_file(Path("catalog.txt"))


def test_extract_catalog_from_zip(tmp_path: Path) -> None:
    """Test extracting catalog from zip file."""
    # Create a test zip file with a catalog
    zip_path = tmp_path / "test_catalog.zip"
    catalog_content = b"fake lightroom catalog content"

    with zipfile.ZipFile(zip_path, "w") as zip_file:
        zip_file.writestr("test_catalog.lrcat", catalog_content)

    # Extract the catalog
    extracted_path = extract_catalog_from_zip(zip_path)

    assert extracted_path.exists()
    assert extracted_path.suffix == ".lrcat"
    assert extracted_path.read_bytes() == catalog_content


def test_extract_catalog_from_zip_no_catalog(tmp_path: Path) -> None:
    """Test extracting catalog from zip file with no .lrcat file."""
    # Create a test zip file without a catalog
    zip_path = tmp_path / "test_no_catalog.zip"

    with zipfile.ZipFile(zip_path, "w") as zip_file:
        zip_file.writestr("some_file.txt", b"not a catalog")

    # Should raise an error
    with pytest.raises(ValueError, match="No .lrcat file found"):
        extract_catalog_from_zip(zip_path)


def test_lightroom_parser_config() -> None:
    """Test LightroomParserConfig default values and validation."""
    # Default config
    config = LightroomParserConfig()
    assert config.ignore_collections == {"quick collection"}
    assert config.include_metadata is True
    assert config.include_keywords is True
    assert config.include_collections is True

    # Custom config
    custom_ignore = {"test collection", "another one"}
    config = LightroomParserConfig(
        ignore_collections=custom_ignore,
        include_metadata=False,
        include_keywords=False,
        include_collections=False,
    )
    assert config.ignore_collections == custom_ignore
    assert config.include_metadata is False
    assert config.include_keywords is False
    assert config.include_collections is False


def test_parse_lightroom_catalog_nonexistent() -> None:
    """Test parsing non-existent catalog file."""
    nonexistent_path = Path("/nonexistent/catalog.lrcat")

    with pytest.raises(FileNotFoundError):
        parse_lightroom_catalog(nonexistent_path)


def test_load_lightroom_catalog_to_df_nonexistent() -> None:
    """Test loading non-existent catalog file with error handling."""
    nonexistent_path = Path("/nonexistent/catalog.lrcat")
    # Should return empty DataFrame with correct schema instead of raising
    with pytest.raises(FileNotFoundError):
        load_lightroom_catalog_to_df(nonexistent_path)


def test_parse_lightroom_catalog_real_data(test_catalogs_dir: Path) -> None:
    """Test parsing real Lightroom catalog files from test data."""
    test_catalogs = [
        test_catalogs_dir / "test_catalog_00" / "test_catalog_fresh.lrcat",
        test_catalogs_dir
        / "test_catalog_01"
        / "test_catalog_gps_captions_collections_keywords.lrcat",
        test_catalogs_dir
        / "test_catalog_02"
        / "test_catalog_two_more_photos_and_edits.lrcat",
        test_catalogs_dir
        / "test_catalog_03"
        / "test_catalog_more_face_tags_gps_edit.lrcat",
    ]

    for catalog_path in test_catalogs:
        # Test with default config
        df = parse_lightroom_catalog(catalog_path)

        # Verify required schema columns exist
        assert SchemaColumns.UUID_STRING in df.columns
        assert SchemaColumns.DATA_TYPE in df.columns
        assert SchemaColumns.NAME in df.columns
        assert SchemaColumns.CAPTION in df.columns
        assert SchemaColumns.SOURCE_FILE in df.columns
        assert SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE in df.columns
        assert SchemaColumns.TIMESTAMP_UTC in df.columns
        assert SchemaColumns.GPS_LATITUDE in df.columns
        assert SchemaColumns.GPS_LONGITUDE in df.columns
        assert SchemaColumns.RATING_0_TO_5 in df.columns
        assert SchemaColumns.SIZE_IN_BYTES in df.columns

        # Check data types
        assert df[SchemaColumns.UUID_STRING].dtype == "object"
        assert df[SchemaColumns.DATA_TYPE].isin([DataType.PHOTO, DataType.VIDEO]).all()

        # Check for valid GPS coordinates if present
        if df[SchemaColumns.GPS_LATITUDE].notna().any():
            non_null_lats = df[SchemaColumns.GPS_LATITUDE].dropna()
            valid_lat = non_null_lats.between(-90, 90, inclusive="both")
            assert (
                valid_lat.all()
            ), f"Invalid latitudes found: {non_null_lats[~valid_lat].tolist()}"

        if df[SchemaColumns.GPS_LONGITUDE].notna().any():
            non_null_lons = df[SchemaColumns.GPS_LONGITUDE].dropna()
            valid_lon = non_null_lons.between(-180, 180, inclusive="both")
            assert (
                valid_lon.all()
            ), f"Invalid longitudes found: {non_null_lons[~valid_lon].tolist()}"

        # Check ratings are in valid range
        if df[SchemaColumns.RATING_0_TO_5].notna().any():
            non_null_ratings = df[SchemaColumns.RATING_0_TO_5].dropna()
            valid_ratings = non_null_ratings.between(0, 5, inclusive="both")
            assert (
                valid_ratings.all()
            ), f"Invalid ratings found: {non_null_ratings[~valid_ratings].tolist()}"


def test_parse_lightroom_catalog_with_metadata(test_catalogs_dir: Path) -> None:
    """Test parsing catalog with metadata enabled."""
    catalog_path = (
        test_catalogs_dir
        / "test_catalog_01"
        / "test_catalog_gps_captions_collections_keywords.lrcat"
    )

    config = LightroomParserConfig(include_metadata=True)
    df = parse_lightroom_catalog(catalog_path, config)

    # Check for metadata columns (updated to match simplified query)
    metadata_columns = [
        "focal_length",
        "aperture",
        "flash_fired",
        "import_hash",
        "file_format",
        "base_name",
        "extension",
        "color_labels",
        "pick",
        "touch_time",
        "original_capture_time",
    ]

    for col in metadata_columns:
        assert col in df.columns, f"Missing metadata column: {col}"


def test_parse_lightroom_catalog_without_metadata(test_catalogs_dir: Path) -> None:
    """Test parsing catalog with metadata disabled."""
    catalog_path = (
        test_catalogs_dir
        / "test_catalog_01"
        / "test_catalog_gps_captions_collections_keywords.lrcat"
    )

    config = LightroomParserConfig(include_metadata=False)
    df = parse_lightroom_catalog(catalog_path, config)

    # Check that extra metadata columns are not present
    metadata_columns = ["focal_length", "aperture", "import_hash", "file_format"]

    for col in metadata_columns:
        assert col not in df.columns, f"Unexpected metadata column found: {col}"


def test_parse_lightroom_catalog_with_keywords(test_catalogs_dir: Path) -> None:
    """Test parsing catalog with keywords enabled."""
    catalog_path = (
        test_catalogs_dir
        / "test_catalog_01"
        / "test_catalog_gps_captions_collections_keywords.lrcat"
    )

    config = LightroomParserConfig(include_keywords=True)
    df = parse_lightroom_catalog(catalog_path, config)

    # Check for keywords column
    assert "keywords" in df.columns

    # Check that keywords are strings when present (converted for SQLite storage)
    keywords_present = df["keywords"].notna()
    if keywords_present.any():
        for keywords in df.loc[keywords_present, "keywords"]:
            assert isinstance(
                keywords, str
            ), f"Keywords should be a string (for SQLite), got {type(keywords)}"


def test_parse_lightroom_catalog_with_collections(test_catalogs_dir: Path) -> None:
    """Test parsing catalog with collections enabled."""
    catalog_path = (
        test_catalogs_dir
        / "test_catalog_01"
        / "test_catalog_gps_captions_collections_keywords.lrcat"
    )

    config = LightroomParserConfig(include_collections=True)
    df = parse_lightroom_catalog(catalog_path, config)

    # Check for collections column
    assert "collections" in df.columns

    # Check that collections are strings when present (converted for SQLite storage)
    collections_present = df["collections"].notna()
    if collections_present.any():
        for collections in df.loc[collections_present, "collections"]:
            assert isinstance(
                collections, str
            ), f"Collections should be a string (for SQLite), got {type(collections)}"


def test_parse_lightroom_catalog_ignore_collections(test_catalogs_dir: Path) -> None:
    """Test parsing catalog with specific collections ignored."""
    catalog_path = (
        test_catalogs_dir
        / "test_catalog_01"
        / "test_catalog_gps_captions_collections_keywords.lrcat"
    )

    # Test with custom ignored collections
    ignore_set = {"test collection", "ignored collection"}
    config = LightroomParserConfig(
        include_collections=True, ignore_collections=ignore_set
    )
    df = parse_lightroom_catalog(catalog_path, config)

    # Check that ignored collections don't appear in results
    if "collections" in df.columns and df["collections"].notna().any():
        all_collections = set()
        for collections_list in df["collections"].dropna():
            if isinstance(collections_list, list):
                all_collections.update(col.lower() for col in collections_list)

        # Check that ignored collections are not present
        for ignored in ignore_set:
            assert (
                ignored.lower() not in all_collections
            ), f"Ignored collection '{ignored}' found in results"


def test_parse_multiple_catalogs(test_catalogs_dir: Path) -> None:
    """Test parsing multiple catalogs and checking for unique UUIDs."""
    test_catalogs = [
        test_catalogs_dir / "test_catalog_00" / "test_catalog_fresh.lrcat",
        test_catalogs_dir
        / "test_catalog_01"
        / "test_catalog_gps_captions_collections_keywords.lrcat",
    ]

    all_dataframes = []
    for catalog_path in test_catalogs:
        df = parse_lightroom_catalog(catalog_path)
        if not df.empty:
            all_dataframes.append(df)

    # Combine all DataFrames
    combined_df = pd.concat(all_dataframes, ignore_index=True)

    # Check for UUID uniqueness across catalogs
    uuid_counts = combined_df[SchemaColumns.UUID_STRING].value_counts()
    duplicates = uuid_counts[uuid_counts > 1]

    # This is informational - we might expect some duplicates between test catalogs
    assert len(all_dataframes) > 0, "Should have parsed at least one catalog"


def test_parse_timezone_aware_timestamps(test_catalogs_dir: Path, snapshot) -> None:
    """Test parsing catalog with various timezone formats in timestamps including NULL values."""
    from syrupy.assertion import SnapshotAssertion
    
    # Use dedicated test catalog with timezone formats and NULL captureTime
    test_catalog = (
        test_catalogs_dir / "test_parse_timezone_aware_timestamps" / "test_parse_timezone_aware_timestamps.lrcat"
    )

    config = LightroomParserConfig(
        include_metadata=False, include_keywords=False, include_collections=False
    )
    df = parse_lightroom_catalog(test_catalog, config)

    # Extract only the timestamp columns for snapshot testing
    timestamp_data = df[
        [
            SchemaColumns.UUID_STRING,
            SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
            SchemaColumns.TIMESTAMP_UTC,
        ]
    ].copy()

    # Convert timezone-aware timestamps to string for snapshot comparison
    # (since timezone-aware timestamps can be tricky to compare in snapshots)
    timestamp_data[SchemaColumns.TIMESTAMP_UTC] = timestamp_data[
        SchemaColumns.TIMESTAMP_UTC
    ].dt.strftime("%Y-%m-%d %H:%M:%S%z")

    # Create snapshot of the timezone handling
    assert timestamp_data.to_dict("records") == snapshot
