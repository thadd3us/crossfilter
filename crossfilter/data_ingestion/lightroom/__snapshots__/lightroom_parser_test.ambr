# serializer version: 1
# name: test_parse_timezone_aware_timestamps
  list([
    dict({
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2007-01-01T16:41:36',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: '2007-01-01 16:41:36+0000',
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>: '23868311-03F0-4F5E-BE52-EDCEE9DEDBE9',
    }),
    dict({
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2020-06-01T01:45:58Z',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: '2020-06-01 01:45:58+0000',
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>: '33F0B4A2-9236-46F2-931C-8413E20B511F',
    }),
    dict({
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: None,
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: nan,
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>: '62E69C3E-A678-486C-AA0E-D287209754CD',
    }),
    dict({
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2020-06-01T01:50:02-07:00',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: '2020-06-01 08:50:02+0000',
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>: '64560483-B9B4-4A77-9E8D-983CB6EB8DD4',
    }),
    dict({
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2020-06-01T01:51:36',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: '2020-06-01 01:51:36+0000',
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>: '8CE5D4D3-AB53-4E82-B73E-3A555DD4D622',
    }),
    dict({
      <SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: 'TIMESTAMP_MAYBE_TIMEZONE_AWARE'>: '2020-06-01T01:49:28+01:00',
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>: '2020-06-01 00:49:28+0000',
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>: 'AD90B5AD-B5E3-4D9A-8234-D9F87A9E40FA',
    }),
  ])
# ---
